src/mgmt/srv_pool.c

```diff
@@ -88,10 +86,122 @@ ds_mgmt_tgt_pool_destroy(uuid_t pool_uuid)
+	map_ranks_fini(excluded);
+	return rc;
+}
+
+/**
+ * Destroy the pool on every DOWN rank
+ */
+static int
+ds_mgmt_tgt_pool_destroy(uuid_t pool_uuid)
+{
+	d_rank_list_t			excluded = { 0 };
+	int				rc;
+
+	rc = ds_pool_get_ranks(pool_uuid, MAP_RANKS_DOWN, &excluded);
+	if (rc)
+		return rc;
+
+	rc = ds_mgmt_tgt_pool_destroy_ranks(pool_uuid, &excluded);
+	if (rc)
+		return rc;
+
+	return DER_SUCCESS;
+}
+
+static int
+ds_mgmt_tgt_pool_create_ranks(uuid_t pool_uuid, char *tgt_dev,
+			      d_rank_list_t *rank_list, size_t scm_size,
+			      size_t nvme_size, uuid_t **tgt_uuids)
+{
+	crt_rpc_t			*tc_req;
+	crt_opcode_t			opc;
+	struct mgmt_tgt_create_in	*tc_in;
+	struct mgmt_tgt_create_out	*tc_out;
+	d_rank_t			*tc_out_ranks;
+	uuid_t				*tc_out_uuids;
+	unsigned int			i;
+	int				topo;
+	int				rc;
+
+	/* Collective RPC to all of targets of the pool */
+	topo = crt_tree_topo(CRT_TREE_KNOMIAL, 4);
+	opc = DAOS_RPC_OPCODE(MGMT_TGT_CREATE, DAOS_MGMT_MODULE,
+			      DAOS_MGMT_VERSION);
+	rc = crt_corpc_req_create(dss_get_module_info()->dmi_ctx, NULL,
+				  rank_list, opc, NULL, NULL,
+				  CRT_RPC_FLAG_FILTER_INVERT, topo, &tc_req);
+	if (rc) {
+		D_ERROR(DF_UUID": corpc_req_create failed: rc="DF_RC"\n",
+			DP_UUID(pool_uuid), DP_RC(rc));
+		return rc;
+	}
+
+	tc_in = crt_req_get(tc_req);
+	D_ASSERT(tc_in != NULL);
+	uuid_copy(tc_in->tc_pool_uuid, pool_uuid);
+	tc_in->tc_tgt_dev = tgt_dev;
+	tc_in->tc_scm_size = scm_size;
+	tc_in->tc_nvme_size = nvme_size;
+	rc = dss_rpc_send(tc_req);
+	if (rc == 0 && DAOS_FAIL_CHECK(DAOS_POOL_CREATE_FAIL_CORPC))
+		rc = -DER_TIMEDOUT;
+	if (rc != 0) {
+		D_ERROR(DF_UUID": dss_rpc_send MGMT_TGT_CREATE: rc="DF_RC"\n",
+			DP_UUID(pool_uuid), DP_RC(rc));
+		D_GOTO(decref, rc);
+	}
+
+	tc_out = crt_reply_get(tc_req);
+	rc = tc_out->tc_rc;
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to update pool map on targets: rc="
+			DF_RC"\n",
+			DP_UUID(tc_in->tc_pool_uuid), DP_RC(rc));
+		D_GOTO(decref, rc);
+	}
+
+	D_DEBUG(DB_MGMT, DF_UUID" create %zu tgts pool\n",
+		DP_UUID(pool_uuid), tc_out->tc_tgt_uuids.ca_count);
+
+	/* Abort early if the caller doesn't need the new pool target UUIDs */
+	if (tgt_uuids == NULL)
+		D_GOTO(decref, rc = DER_SUCCESS);
+
+	/* Gather target uuids ranks from collective RPC to start pool svc. */
+	D_ALLOC_ARRAY(*tgt_uuids, rank_list->rl_nr);
+	if (*tgt_uuids == NULL) {
+		rc = -DER_NOMEM;
+		D_GOTO(decref, rc);
+	}
+	tc_out_ranks = tc_out->tc_ranks.ca_arrays;
+	tc_out_uuids = tc_out->tc_tgt_uuids.ca_arrays;
+	for (i = 0; i < tc_out->tc_tgt_uuids.ca_count; i++) {
+		int	idx;
+		bool	found;
+
+		found = daos_rank_list_find(rank_list, tc_out_ranks[i], &idx);
+		D_ASSERT(found);
+
+		/* copy returned target UUID */
+		uuid_copy((*tgt_uuids)[idx], tc_out_uuids[i]);
+
+		D_DEBUG(DB_TRACE, "fill ranks %d idx %d "DF_UUID"\n",
+			tc_out_ranks[i], idx, DP_UUID(tc_out_uuids[i]));
+	}
+
+	rc = DER_SUCCESS;
+
+decref:
+	crt_req_decref(tc_req);
+	if (rc)
+		ds_mgmt_tgt_pool_destroy(pool_uuid);
```

Comment:
Should only do the destroy on specified ranks, right? Could be fixed later when we do real pool creation on reintegration.

Reply:
Good catch - changed this to just the specific ranks.